src/viam/ur/module/ur_arm.cpp

```diff
@@ -172,128 +184,1284 @@ void write_waypoints_to_csv(const std::string& filepath, const std::list<Eigen::
+    std::exchange(cancellation_request, {})->promise.set_exception(std::make_exception_ptr(std::runtime_error{std::string{message}}));
+}
+
+void URArm::state_::move_request::write_joint_data(vector6d_t& position, vector6d_t& velocity) {
+    ::write_joint_data(position, velocity, arm_joint_positions_stream, unix_time_iso8601(), arm_joint_positions_sample++);
+}
+
+URArm::state_::move_request::cancellation_request::cancellation_request() {}
+
+void URArm::state_::emit_event_(event_variant_&& event) {
+    std::visit([this](auto&& event) { this->emit_event_(std::forward<decltype(event)>(event)); }, std::move(event));
+}
 
-    // variables specified by ResourceConfig and set through reconfigure
-    std::string host;
-    std::atomic<double> speed{0};
-    std::atomic<double> acceleration{0};
-    std::atomic<bool> estop{false};
-    std::atomic<bool> local_disconnect{false};
+template <typename T>
+void URArm::state_::emit_event_(T&& event) {
+    current_state_ = std::visit(
+        [&](auto& current_state) {
+            const auto current_state_description = current_state.describe();
+            const auto event_description = event.describe();
+            auto new_state = current_state.handle_event(std::forward<T>(event));
+            const auto new_state_description = std::visit([](const auto& state) { return std::string{state.describe()}; }, new_state);
+            VIAM_SDK_LOG(info) << "URArm state transition from state `" << current_state_description << "` to state `"
+                               << new_state_description << "` due to event `" << event_description << "`";
+            return new_state;
+        },
+        current_state_);
+}
+
+std::chrono::milliseconds URArm::state_::get_timeout_() const {
+    return std::visit([](auto& state) { return state.get_timeout(); }, current_state_);
+}
+
+std::string URArm::state_::describe_() const {
+    return std::visit([](auto& state) { return state.describe(); }, current_state_);
+}
+
+void URArm::state_::upgrade_downgrade_() {
+    if (auto event = std::visit([this](auto& state) { return state.upgrade_downgrade(*this); }, current_state_)) {
+        emit_event_(*std::move(event));
+    }
+}
+
+void URArm::state_::clear_ephemeral_values_() {
+    ephemeral_.reset();
+}
+
+void URArm::state_::recv_arm_data_() {
+    if (auto event = std::visit([this](auto& state) { return state.recv_arm_data(*this); }, current_state_)) {
+        emit_event_(*std::move(event));
+    }
+}
 
-    // variables derived from arm state on configuration
-    bool is_sim;
+void URArm::state_::handle_move_request_() {
+    if (auto event = std::visit([this](auto& state) { return state.handle_move_request(*this); }, current_state_)) {
+        emit_event_(*std::move(event));
+    }
+}
+
+void URArm::state_::send_noop_() {
+    if (auto event = std::visit([](auto& state) { return state.send_noop(); }, current_state_)) {
+        emit_event_(*std::move(event));
+    }
+}
+
+void URArm::state_::run_() {
```

Comment:
The high-frequency loop. We get the lock, use a condvar to wait either for the per-state timeout or the exit signal. If we wake up due to timeout, then it is time to drive the state machine. We call each activity in order. As we advance through the methods, the internal state may change method to method.